File: getting-started-mocks/02-first-useful-mock/python/before/notifier.py
"""The package notifier before it had a seam.

It talks to the carrier directly, which is exactly why it cannot be tested:
nothing in a test can make the carrier say a package is delayed.
"""

import json
from urllib.request import urlopen

CARRIER_URL = "https://api.example-carrier.test"


def notify(tracking_number: str) -> str | None:
    """Return the message to send to the customer, or None when there is nothing to say."""
    with urlopen(f"{CARRIER_URL}/shipments/{tracking_number}") as response:
        shipment = json.load(response)

    if shipment["status"] == "delayed":
        return f"Package {tracking_number} is delayed"
    return None

File: getting-started-mocks/02-first-useful-mock/node/before/notifier.js
// The package notifier before it had a seam. It talks to the carrier directly,
// which is exactly why it cannot be tested: nothing in a test can make the
// carrier say a package is delayed.

const carrierUrl = 'https://api.example-carrier.test'

// Returns the message to send to the customer, or null when there is nothing to say.
export async function notify(trackingNumber) {
  const response = await fetch(`${carrierUrl}/shipments/${trackingNumber}`)
  const shipment = await response.json()

  if (shipment.status === 'delayed') {
    return `Package ${trackingNumber} is delayed`
  }
  return null
}
